### ISSUE TITLE
Question for handling different size of input via masking

### ISSUE THREAD

**Issue by Nanco-L** [e3nn/e3nn#104]
Hello.

I want to use the molecule/crystal training data.
As you already know, the number of atoms in molecules is not fixed.
And common method to generate batch from a dataset like this is masking.
Thus, I need to handle a dataset like below:
```python
features=[ # [2 (batch size), 3 (# of atoms), # of features]
    [
        [...],
        [...],
        [empty space] # masked with some value
    ],
   [
        [...],
        [...],
        [...] # masked
   ]
]
``` 
There are two atoms in first molecule and three atoms in second molecule.
And what I want is generating model that provide same output regardless of the masked space. 
Is it possible in e3nn module?
Please guide me for the question, thanks.

**Comment by blondegeek** [comment]
Hi @Nanco-L,

You'll want to use our Convolutions that use `torch_geometric` data types for batching. To use these Convolutions, you will need to process your data to have neighborlists (edges) and relative distance vectors as edge features with the classes in e3nn.point.data_helpers. Please see this tutorial for the relevant data construction classes which inherit from `torch_geometric.data.Data`:
https://github.com/blondegeek/e3nn_tutorial/blob/master/datatypes_for_neighbors.ipynb

As for the network itself, (since I have yet to make the tutorial for this feature) I recommend using the following style of network at least to start:
```
import e3nn
import torch_geometric
from e3nn.networks import GatedConvParityNetwork
from e3nn.point.message_passing import Convolution

# You will need to define Rs_in, Rs_out according to your task
# Note for GatedConvParityNetwork you will need to use an Rs that specifies [(multiplicity, L, parity)]
# e.g if the inputs are scalars and outputs are vectors
# Rs_in = [(num_scalars, 0, 1)].  # Scalars do not change sign under inversion.
# Rs_out = [(num_vectors, 1, -1)]. # Vectors do change sign under inversion.

model_kwargs = {
    "convolution": Convolution,  # Note this specifically uses the Convolution based on torch_geometric.nn.MessagePassing
    "Rs_in": Rs_in,
    "Rs_out": Rs_out,
    "mul": 4,  # This is essentially the number of channels per "L". Start small and go up from there.
    "layers": 3,
    "max_radius": 5.0,  # Should match whatever you used in DataNeighbors or DataPeriodicNeighbors
    "lmax": 2,  # lmax = 2 should be sufficient to start
    "number_of_basis": 10 
}

model = GatedConvParityNetwork(**model_kwargs)

# define data as a list of DataNeighbors or DataPeriodicNeighbors 
# See tutorial for how to do this

dataloader = torch_geometric.data.DataLoader(data, batch_size=batch_size)
n_norm = 30  # Change this to the average number of nearest neighbors per node (over entire training set)
opt = torch.optim.Adam(model.parameters(), 3e-3)

# Train example
for j, d in enumerate(dataloader):
    output = model(d.x, d.edge_index, d.edge_attr, n_norm=n_norm, batch=d.batch)
    loss = loss_fn(output, d.y)
    opt.zero_grad()
    loss.backward()
    opt.step()
```

Hopefully that helps! Please feel free to ask further questions.

**Comment by Nanco-L** [comment]
Thank you for kind answer, @blondegeek.

However, what I exactly want to know is, is it possible to create batch using molecules consist of different number of atoms?
The concept is similar with NLP. In the NLP model, batch contains multiple sentences consist of different number of words.

For example, I want to make batch with H2O and CH4. Then, dimension of features(`feature_1`) and geometry(`geometry_1`) (input for point convolution in `e3nn/point/operations.py`) are `[2, 5, # of features]` and `[2, 5, 3]`, respectively.
In this case, `features_1[:,2:,:]` and `geometry_1[:,2:,:]` are meaningless space.
In other case, H2O and CH3COOH can be included in same batch and `feature_2.size() == (2, 8, # of features)` and `geometry_2.size() == (2, 8, 3)`.
But `features_2[:,2:,:]` and `geometry_2[:,2:,:]` are still meaningless space.
And even if the length of the first and second batches are different, `model(feature_1, geometry_1)[0] == model(feature_2, geometry_2)[0]` since it contains same molecules(H2O).

Please guide the method to create the model can use batch that I explain above.
Or, if that is already possible with `torch_geometric` data type, please notice me.
Thanks!

**Comment by blondegeek** [comment]
Batching molecules with different numbers of atoms is exactly what the
DataNeighbor, message_passing.Convolution, and the torch geometric
DataLoader is built to do. It does involve preprocessing — constructing the
neighborhood graph. The reason for this is that batches are essentially
“big graphs” where edges only go between nodes (atoms) in the same example.
You do not have a batch dimension but rather a batch array that indicates
which example each node belongs to. Take a look at the torch_geometric
documentation for more information on their data structures which ours
inherit from.

**Comment by Nanco-L** [comment]
Thanks for fast response!
I'll check the document and test it.
